src/library/scala/collection/mutable/ArrayBuilder.scala

```diff
@@ -45,14 +45,16 @@ sealed abstract class ArrayBuilder[T]
 
   protected[this] def resize(size: Int): Unit
 
-  /** Add all elements of an array */
+  /** Add all elements of an array. */
   def addAll(xs: Array[_ <: T]): this.type = addAll(xs, 0, xs.length)
 
-  /** Add a slice of an array */
+  /** Add a slice of an array. */
   def addAll(xs: Array[_ <: T], offset: Int, length: Int): this.type = {
-    ensureSize(this.size + length)
-    Array.copy(xs, offset, elems, this.size, length)
-    size += length
+    if (length > 0) {
```

Comment:
if (length - offset > 1 && xs.isNotEmpty)

Reply:
I see, instead of `ensureSize` for some `length`, first make sure the capacity is needed.

xs.length != 0 && xs.length - offset > 0 and the effective length is min(length, xs.length - offset)

I think all this was deferred to `Array.copy` for convenience. But the downside of erroneously large length makes it worth guarding here.